syncer/syncer.go

```diff
@@ -1318,9 +1326,53 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		failpoint.Inject("ProcessBinlogSlowDown", nil)
 
 		s.tctx.L().Debug("receive binlog event", zap.Reflect("header", e.Header))
+
+		startLocation = currentLocation.Clone()
+		// TODO: support all event
+		// we calculate endLocation(currentLocation) for Rows/Query event here
+		switch e.Event.(type) {
+		case *replication.RowsEvent, *replication.QueryEvent:
+			endSuffix := startLocation.Suffix
+			if s.firstReplace {
+				endSuffix = 1
+				s.firstReplace = false
+			} else if endSuffix > 0 {
+				endSuffix++
+			}
+
+			currentLocation = binlog.Location{
+				Position: mysql.Position{
+					Name: lastLocation.Position.Name,
+					Pos:  e.Header.LogPos,
+				},
+				GTIDSet: lastLocation.GTIDSet.Clone(),
+				Suffix:  endSuffix,
+			}
+			if ev, ok := e.Event.(*replication.QueryEvent); ok {
+				currentLocation.GTIDSet.Set(ev.GSet)
+			}
+
+			apply, op := s.errOperatorHolder.Apply(&startLocation, &currentLocation)
```

Comment:
could we move this `Apply` earlier? for example, below `e, err = s.getEvent(tctx, &currentLocation)`. 

and if user-specified `pos` is at wrong position, should we log something when `RemoveOutdated` removes it?

Reply:
> log something when RemoveOutdated removes it?

add log in 92d933c

> could we move this Apply earlier?

But we still handle the error for getEvent before.😥